consensus/src/liveness/leader_reputation.rs

```diff
@@ -215,24 +210,24 @@ impl NewBlockEventAggregation {
 
         (&history[start..])
             .iter()
-            .filter(move |&meta| meta.epoch() == epoch)
+            .filter(move |&meta| epoch_to_candidates.contains_key(&meta.epoch()))
     }
 
     pub fn get_aggregated_metrics(
         &self,
         epoch: u64,
-        candidates: &[Author],
+        epoch_to_candidates: &HashMap<u64, Vec<Author>>,
         history: &[NewBlockEvent],
     ) -> (
         HashMap<Author, u32>,
         HashMap<Author, u32>,
         HashMap<Author, u32>,
     ) {
-        let votes = self.count_votes(epoch, candidates, history);
-        let proposals = self.count_proposals(epoch, history);
-        let failed_proposals = self.count_failed_proposals(epoch, candidates, history);
+        let votes = self.count_votes(epoch_to_candidates, history);
+        let proposals = self.count_proposals(epoch_to_candidates, history);
+        let failed_proposals = self.count_failed_proposals(epoch_to_candidates, history);
 
-        for candidate in candidates {
+        for candidate in &epoch_to_candidates[&epoch] {
```

Comment:
I just realize this actually dumps metrics for all nodes, is it a lot for every round?

Reply:
@sitalkedia who added this. It is a guage metric per node, not reported on every round, I think.

Reply:
@zekun000 - This is super useful to get a view of all the validator's performance in real-time until we have prometheus metrics collection support from third party validator. Once we have that, we can change it back to emitting metrics only for this node.